ext/openssl/ossl_x509cert.c

```diff
@@ -704,6 +704,94 @@ ossl_x509_eq(VALUE self, VALUE other)
     return !X509_cmp(a, b) ? Qtrue : Qfalse;
 }
 
+static VALUE
+ossl_x509_load_chained_cert_from_file(VALUE self, VALUE path)
+{
+    VALUE certs; /* Return array (Certificate object) */
+    BIO *in;
+    X509 *x509, *ca;
+    STACK_OF(X509) *certs_stack;
+    unsigned long err;
+    SSL_CTX *ctx = SSL_CTX_new(TLS_method());
+
+    in = BIO_new(BIO_s_file());
+    if (in == NULL) {
+        SSL_CTX_free(ctx);
+        ossl_raise(eX509CertError, NULL);
+    }
+
+    if (BIO_read_filename(in, StringValueCStr(path)) <= 0) {
+        SSL_CTX_free(ctx);
+        ossl_raise(eX509CertError, NULL);
+    }
+
+    x509 = PEM_read_bio_X509_AUX(in, NULL, NULL, NULL);
+    if (x509 == NULL) {
+	BIO_free(in);
+        SSL_CTX_free(ctx);
+        ossl_raise(eX509CertError, NULL);
+    }
+
+    SSL_CTX_use_certificate(ctx, x509);
+    if (ERR_peek_error() != 0) {
+        X509_free(x509);
+	BIO_free(in);
+        SSL_CTX_free(ctx);
+        ossl_raise(eX509CertError, NULL); /* Key/certificate mismatch doesn't imply */
+    }
+
+    /*
+     * If we could set up our certificate, now proceed to the CA
+     * certificates.
+     */
+    SSL_CTX_clear_chain_certs(ctx);
+    while ((ca = PEM_read_bio_X509(in, NULL, NULL, NULL)) != NULL) {
+        int r = SSL_CTX_add0_chain_cert(ctx, ca);
+        /*
+         * Note that we must not free ca if it was successfully added to
+         * the chain (while we must free the main certificate, since its
+         * reference count is increased by SSL_CTX_use_certificate).
+         */
+
+        if (!r) {
+            X509_free(ca);
+            X509_free(x509);
+            BIO_free(in);
+            SSL_CTX_free(ctx);
+            ossl_raise(eX509CertError, NULL);
+        }
+    }
+
+    /* When the while loop ends, it's usually just EOF. */
+    err = ERR_peek_last_error();
+    if (! (ERR_GET_LIB(err) == ERR_LIB_PEM
+        && ERR_GET_REASON(err) == PEM_R_NO_START_LINE)) {
+        /* some real error */
+        X509_free(x509);
+        BIO_free(in);
+        SSL_CTX_free(ctx);
+        ossl_raise(eX509CertError, NULL);
+    }
+
+    ERR_clear_error();
+
+    certs = rb_ary_new();
+    SSL_CTX_get0_chain_certs(ctx, &certs_stack);
+
+    for (int cert_idx = 0; cert_idx < sk_X509_num(certs_stack); ++cert_idx) {
+        X509 * cert_c = sk_X509_value(certs_stack, cert_idx);
+        VALUE cert = ossl_x509_new(cert_c);
+        rb_ary_push(certs, cert);
```

Comment:
(Although it's kind of rare condition,) rb_ary_new(), ossl_x509_new(), and rb_ary_push() involve memory allocation and can potentially raise a Ruby exception on failure, which is basically a longjmp().

To prevent possible memory leak, these function calls have to be wrapped by rb_protect().

Reply:
Thank you for careful review!

I think I found the proper supplement[3] for this potential Ruby exception with memory leak problem. I'll try apply `rb_protect()` soon.

[3] https://silverhammermba.github.io/emberb/c/#rb_protect